### ISSUE TITLE
[RFC] Add Columns/Column component for vertical layout

### ISSUE THREAD

**Issue by afontcu** [holaluz/margarita#274]
Related to #256.

## Impact

- New `MaColumns` component.
- New `MaColumn` component.
- Deprecate `MaGridContainer`,  `MaGridRow`, `MaGridColumn`.

## Details

This RFC aims to include two-component set, with a simpler API, which can be used in conjunction with other layout components (Stack and other) and the responsive props proposed in #273.

## Proposed API

```vue
<ma-columns>
  <ma-column>content 1</ma-column>
  <ma-column>content 2</ma-column>
</ma-columns>
```

### Props for `MaColumns`

```js
space // layout-based space modifiers (small, medium, large…). Same as Stack.
collapse-below // responsive values (mobile, tablet, desktop)
align // align modifiers (start, center, left). Same as Stack. Useful in combination with collapse-bellow, and when columns do not reach 100% width.
align-vertical: "top" / "bottom" / "center". // Responsive prop (can receive an array of values)
```

`collapse-bellow` converts a column into a stack below the specified breakpoint. We could somehow replace MaColumn with MaStack (via `<component is="">`), but I believe this is such a common scenario it makes sense to provide a prop to solve this specific use case.

### Props for `MaColumn`

```js
width: "content" / "1/2" / "1/3" / "2/3" / "1/4" / "3/4"
```

If not specified, columns' width will be fluid, meaning they will take up to all available space:

```vue
<ma-columns>
  <ma-column width="1/3">...</ma-column>
  <ma-column>....</ma-column> <!-- it'll take two thirds of space -->
</ma-columns>
```

```vue
<ma-columns>
  <ma-column width="3/4">...</ma-column>

  <!-- both columns will split the remaining space (1/4) equally -->
  <ma-column>...</ma-column>
  <ma-column>....</ma-column>
</ma-columns>
```

`"content"` means a column that only takes as much space as needed:

```vue
<ma-columns>
  <ma-column width="content"><div style="height:30px;width:80px"></div></ma-column> <!-- 80px -->
  <ma-column>....</ma-column> <!-- 100% of width minus 80px -->
</ma-columns>
```

(Do we need fifths?)

## Examples

Collapsed below tablet with responsive space. Small gap on mobile (also collapsed), large otherwise.

Also, second column takes up two thirds of space.

```vue
<ma-columns collapse-below="tablet" :space="['small', 'large']">
  <ma-column>...</ma-column>
  <ma-column width="2/3">...</ma-column>
</ma-columns>
```

Columns width default (`medium`) spacing that become horizontally centered on mobile:

```vue
<ma-columns collapse-below="tablet" align="center">
  <ma-column>...<ma-column>
  <ma-column>...<ma-column>
  <ma-column>...<ma-column>
</ma-columns>
```

## Alternatives

Not developing this component at all, an rely on CSS Grid + layout space tokens to create our columns. We'd miss responsive props, but we'd win several degrees of freedom – not sure if a good or a bad thing XD

---

Thoughts? Ideas? Questions? Doubts?

**Comment by sergigraciaj** [comment]
A codesandbox to see theory in action as u did for the Stack would help me a lot! ^^
I like the concept overall, for now I would just say to look for another name on `ma-columns` component, it's to similar to `ma-column`, but I'm not sure... container, wrapper, grid... something with 'horizontal' before? lol

**Comment by afontcu** [comment]
> for now I would just say to look for another name on `ma-columns` component, it's to similar to `ma-column`, but I'm not sure... container, wrapper, grid... something with 'horizontal' before?

hm, true, even though I strongly believe we should deprecate `MaGrid*` components, possibly by moving them to projects that are using them (Bonasera and… that's it??) instead of having them available in Margarita 😉

**Comment by afontcu** [comment]
> A codesandbox to see theory in action as u did for the Stack would help me a lot! ^^

Sure! Fancy to work on it? 😏

**Comment by sergigraciaj** [comment]
> > A codesandbox to see theory in action as u did for the Stack would help me a lot! ^^
> 
> Sure! Fancy to work on it? 😏

Let's give it a go! But I will need ur help, master!

**Comment by afontcu** [comment]
Any progress on this? 🙌

**Comment by sergigraciaj** [comment]
Still WIP, but https://codesandbox.io/s/still-sea-tpkwj?file=/src/App.vue
- For the wrapper component I use "row" instead of "columns" for the name
- Using flex, with margin for the spacing
- For the column's width, I dit it in %, is the best approach?
- For the columns I used a "base 12 grid" in where you just specify how columns is using. If you use 3 of the 12 columns, u just add a prop _width="3"._ If that's not clear enough it can be _width="3/12"_ .
- Not sure about this base 12 thing, probably we need to agree this with design people.
- I still got to add the full width and content width.
- I still got to find how to solve the responsive

**Comment by afontcu** [comment]
> A codesandbox to see theory in action as u did for the Stack would help me a lot! ^^
> I like the concept overall, for now I would just say to look for another name on `ma-columns` component, it's to similar to `ma-column`, but I'm not sure... container, wrapper, grid... something with 'horizontal' before? lol

(Just reread this, lol) That's the whole point, right? `MaColumns` would only be a wrapper aimed to hold `MaColumn`, so I think that making the relationship explicit is beneficial

**Comment by Samitier** [comment]
I have some fresh knowledge about this after doing the `ma-grid-` refactors, so I'm going to give my 5 cents:

I think this part of the UI is a pretty good use case for deciding an API. I will base my examples on this:

![image](https://user-images.githubusercontent.com/4160121/104583436-66b6bb00-5661-11eb-8d8f-7e429810d1b1.png)

So here are my ideas:

### Why 2 components?

I think that to be consistent with `ma-stack`, we should only have one parent component. IMHO it will be way clearer this way:

```html
<ma-columns cols="3/12 5/12 4/12">
  <div>Código postal...</div>
  <div>Localidad...</div>
  <div>Provincia...</div>
</ma-columns>
```

### Now let's make it more similar to CSS grids:

What if we use [`grid-template-columns` and `fr` units](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-columns) as an inspiration for the prop that decides the size of the rows? For example:

```html
<ma-columns cols="1 3 2">
  <div>Código postal...</div>
  <div>Localidad...</div>
  <div>Provincia...</div>
</ma-columns>
```
Meaning that the first col will occupy only 1/6 of the space, the second 3/6 and the third 2/6.

### Soo...is this component really going to control horizontal space or also vertical?

Since we are basing the implementation on css grids, what if we open the possibility to add multiple rows too. Instead of doing that with a stack like:

```html
<ma-stack>
   <ma-columns>
       row1
    </ma-columns>
    <ma-columns>
       row2
    </ma-columns>
</ma-stack>
```

We could do something like this for all the immediate childs of ma-grid:

```html
<ma-grid columns="1 1 1">
   div (row1)
   div (row1)
   div (row1)
   div (row2)
   div (row2)
   div (row2)
   ...
</ma-grid>
```

And even defining different sizes per row:

```html
<ma-grid columns="[ '1 2 3', '3 2 1' ]">
   ...
</ma-grid>
```

### What about spacing between elements?

With CSS grids we can control both horizontal spacing and vertical spacing, so implementing this would be trivial:

```html
<ma-grid gap="sm">
or
<ma-grid gap-col="sm">
or
<ma-grid gap-row="sm">
```

### What about responsiveness?

We could have multiple props for responsiveness. Example: we have a grid but we want a single div stack for mobile:

```html
<ma-grid columns="[ '1 2 3', '3 2 1' ]" columns-mobile="1">
```

### Final example:

So, let's implement the example from the image:

```html
<ma-grid columns="['1 3 2' , '2 3 1', '1 1 1 1']" columns-mobile="1" gap="sm">
  <div>Código postal...</div>
  <div>Localidad...</div>
  <div>Provincia...</div>
  <div>Tipo de via</div>
  <div>Nombre de la via</div>
  <div>Numero</div>
  <div>Portal</div>
  <div>Escalera</div>
  <div>Piso</div>
  <div>Puerta</div>
</ma-grid>
```

### Final thoughts

I like an approach like this because it will be very similar to CSS Grids, so we can use css grids as a baseline for deciding which features to add or not.

But, there are stuff that I don't like too: 

  - What happens with ma-stack then? Will it become redundant?
  - What happens when the dom is dynamic (when there are v-ifs). Should we have to change the columns prop too? IMHO it should do exactly the same as css grids.
  - How we restrict the design system requirement of 12 columns per row with this? Aren't we giving too much freedom here?
  
  Thoughts?

**Comment by afontcu** [comment]
Hey! That's a sweet proposal 👌 

Here's a recap after a quick call we made:

* Both components (MaColumn(s) and MaGrid) make sense, because they cover slightly different use cases.
* MaColumn(s) and MaStack lay out components in a single dimension, whereas MaGrid handles a component matrix.
* Thus, MaGrid API could be tailored to this use case: to provide a grid.
* …thus, its usage would be limited to those cases where the layout is really complex, and achieving the same result with columns and stack would yield a lot of boilerplate.
 
That being said, we still need to know if our design system should allow for an indeterminate number of columns, or if we should have the usual 12-col constraint. Let's think in terms of every team&developer, not only us 😉 

About the current API, I found it a bit confusing with the responsive props we're using in MaStack (`:space=['small', 'large']`). 

_@Samitier feel free correct me if I misunderstood anything!_

**Comment by Samitier** [comment]
Hello, I've been thinking about this and there is stuff that bothers me a bit:

### 1 - Why diferentiating between stack and column, vertical and horizontal?

Ok, so what I'm trying to say is: why are we making this distinction exactly? Isn't the line a lot blurrier than this when we think of a layout? 

For example: an hypothetical responsive column layout will inevitably become a stack for mobile when the columns don't fit anymore. Isn't it a bit weird? We have stacks and columns and one can become another depending of the size of the screen... why do we need to name them differently then, what is the purpose of it besides adding complexity to the code we write?

### 2 - Our layouts will ALWAYS be a composition of stacks and columns
No matter how we implement this, we will always have this dependency. All of our layouts will always be a composition of stack and cols. So why don't we embrace it? Why do we have to waste time thinking we have to implement one or another? Why don't we just let another component do the work for us and just tell it how we want the layout via props?

### 3 - Introducing the `ma-layout` component
Crazy idea time. Let's stop thinking if a ux is stack or col for a second and assume it's simply a layout that we can reorder depending of the screen size. This ma-layout component hides the complexity by instantiating stacks and cols for us depending on the props we pass. 

For example a stack:
```html
<ma-layout gap="sm">
   <div>Row 1</div>
   <div>Row 2</div>
   <div>Row 3</div>
</ma-layout>

Same as this:
<ma-layout gap="sm" cols="1">
   <div>Row 1</div>
   <div>Row 2</div>
   <div>Row 3</div>
</ma-layout>

```

A column layout of constant sizes:
```html
<ma-layout gap="sm" cols="1 1 1">
   <div>Col 1</div>
   <div>Col 2</div>
   <div>Col 3</div>
</ma-layout>
```

A composition of columns and stacks of constant size:
```html
<ma-layout gap="sm" cols="1 1 1">
   <div>Row1 Col1</div>
   <div>Row1 Col2</div>
   <div>Row1 Col3</div>
   <div>Row2 Col1</div>
   <div>Row2 Col2</div>
   <div>Row2 Col3</div>
</ma-layout>
```
A composition of columns and stacks of different sizes per row and col:
```html
<ma-layout gap="sm" cols="3 1 2 - 1 3 2">
   <div>Row1 Col1</div>
   <div>Row1 Col2</div>
   <div>Row1 Col3</div>
   <div>Row2 Col1</div>
   <div>Row2 Col2</div>
   <div>Row2 Col3</div>
</ma-layout>
```

A composition of columns and stacks that become a stack for mobile:
```html
<ma-layout gap="sm" cols="1 1 1" cols-mobile="1">
   <div>Row1 Col1</div>
   <div>Row1 Col2</div>
   <div>Row1 Col3</div>
   <div>Row2 Col1</div>
   <div>Row2 Col2</div>
   <div>Row2 Col3</div>
</ma-layout>
```

The implementation is pretty trivial too, by using CSS grids and the grid-template-columns prop.

**TLDR:**  what if we make `ma-stack` and `ma-column` "private" components that are only rendered by an abstaction called `ma-layout`. This component instantiates one or another depending on responsiveness and the props we send it.

What do you think?

**Comment by afontcu** [comment]
I like the idea! Let's explore it. The migration path from `MaStack` looks quite straightforward. 

I'd like to iterate the API a bit:

1) The responsive props (`cols-mobile`) might be comfortable, but is deeply coupled to our _current_ breakpoints.
2) What about other units? I assume `cols="1 1 1"` means `1fr 1fr 1fr`, but could I also set `cols="150px min-content 2fr"`? 
3) We'd need to handle both `gap-col` and `gap-row`. Also I would make them match our current spacing naming (small instead of sm).

Also, how can I create a list of columns, that is, a variable list of horizontally-aligned items? Maybe we could add a prop called `flow` (to match [grid-auto-flow](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-auto-flow)):

```vue
<ma-layout flow="col">
  <div>[1]</div>
  <div>[2]</div>
  <div>[3]</div>
  <div>[4]</div>
</ma-layout> 
```

would yield 

```
[1] [2] [3] [4]
```

---

A question remains, though: Is the DS going to set some sort of limitations when it comes to the number of columns?

**Comment by edufarre** [comment]
I believe it would be awesome to have everything kinda wrapped up inside one unique component, a.k.a `ma-layout`. 
However, as @afontcu already said, I am not sure about having a prop like `cols-mobile` , being honest, I do think that having a prop where you could pass a 'template' for the grid for each of the breakpoints would be awesome (bear in mind the way that `ma-stack` is currently behaving with the prop `align` when its value is an `array` that relates the values with the different `breakpoint` values.

**Comment by Samitier** [comment]
> * The responsive props (`cols-mobile`) might be comfortable, but is deeply coupled to our _current_ breakpoints.

Yes, completely agree with you and @edufarre . The examples were written to express the idea clearly, but we should adapt the API to what we have in ma-stack, for example making `cols` and `gap` accept an array.

> * What about other units? I assume `cols="1 1 1"` means `1fr 1fr 1fr`, but could I also set `cols="150px min-content 2fr"`?  The original question still remains: Is the DS going to set some sort of limitations when it comes to the number of columns?

That's a good question. I would love some kind of "restriction" here in the DS in order to not get too "creative" when designing pages 😂 . I mean not directly exposing the `grid-template-columns` prop, but having some rules, like sticking to 12 col layouts as @edufarre proposes. 

I personally like the `fr` units because we can design "fluidly" without having to rely in 12 columns or fixed widths but we should think about what we really need that makes us all confortable. The UXrs could really help us here! 😃 

> 3\. We'd need to handle both `gap-col` and `gap-row`. Also I would make them match our current spacing naming (small instead of sm).

Aggree 100%. And make them both handle responsiveness too.

> Also, how can I create a list of columns, that is, a variable list of horizontally-aligned items? Maybe we could add a prop called `flow` (to match [grid-auto-flow](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-auto-flow)):

I thought about "playing" with the cols prop. Maybe `cols="*"` for layouts with "infinite" cols, so that we can do complex layouts with it too. For example, a stack with one full width cols and 15 cols below:
```html
<ma-layout cols="1 - *">
    <div>Full width row</div>
    <div v-for="i in 15">col {{ i }}</div>
</ma-layout>
```

But yeah, maybe this `col` prop already is doing too many "magic", I don't know what to think... 😅

**Comment by Samitier** [comment]
Other stuff we can discuss:
 - Alignment. Should this component care about alignment at all or rely on `align-self` & `justify-self` in the child divs? I vote for the latter because this component will have way less responsibilities/complexity. I can't imagine a simple way to center an entire grid with only one prop 😂 
 - Ordering: same as the avobe, maybe the children should do it with css.
 - Handling "Empty" columns. Imagine a layout like this:
 
![image](https://user-images.githubusercontent.com/4160121/105144730-f7781580-5afd-11eb-8595-5a4d7531b451.png)

How do we leave this space "empty"? With the api we discussed we can simply leave an empty div there and it will render like that. Do you think its a good idea or a bit hacky to have empty divs?

By the way @edufarre, a "nice" way to have 12 column layouts with `fr` units that I didn't think of yesterday: simply make sure that you put values that add up to 12 for each row. For example:

`cols="7 5"` is the same as `<div class="col-7" /> <div class="col-5" />`. I have done the 12 column layout in the codesandbox if you want to see it.

**Comment by afontcu** [comment]
^ several of the questions above lead me to the following:

```vue
<ma-layout>
  <!-- matches `grid-column: 3`, that is, places element after in the third column cell  -->
  <ma-layout-item column="3">...</ma-layout-item>

  <!-- it fills two columns -->
  <ma-layout-item span="2">...</ma-layout-item> 

  <!-- same API as Stack -->
  <ma-layout-item :align="['center', 'right']">...</ma-layout-item>
</ma-layout>
```

😄

**Comment by Samitier** [comment]
> ^ several of the questions above lead me to the following:

Yes looks good to me, defining alignments and such directly on the child components. I don't know the freedom we should give tho. 🤔 

> Related this code: cols="3 1 2 - 1 3 2". For me, visually , is a little bit hard to understand. I would rather to define the cols just with one number in the element (being 12 the maximum).

Yes, the problem with that is that we will still separate our DOM between "rows" and "cols", which is a bit false if, for example, we have different layouts for mobile/desktop (its the same stuff we discussed with stack and column). Also I don't know who would be responsible for the vertical gaps in there.

> 2. Using cols="3 1 2 - 1 3 2" . I'm not convinced to use empty elements (divs) for empty cols.

Same 😂 . Smells a bit. Maybe we could "play"  with the cols prop too. For example: `.` for marking empty divs or something like that. Again, maybe we are overloading this prop with responsibility or we are making ilegible code at this point...

Example for an empty grid `cols="1 3. - *"` will render a col of size 1, followed by an empty space of size 3, followed by another row with "infinite" columns.
